src/python/pants/backend/javascript/nodejs_project.py

```diff
@@ -78,6 +80,10 @@ def generate_lockfile_args(self) -> tuple[str, ...]:
     def immutable_install_args(self) -> tuple[str, ...]:
         if self.package_manager == "npm":
             return ("clean-install",)
+        if self.package_manager == "yarn":
+            if nodesemver.satisfies(self.package_manager_version, "1.x"):
+                return ("install", "--frozen-lockfile")
+            return ("install", "--immutable")
         return ("install", "--frozen-lockfile")
```

Comment:
The trailing return here doesn't make sense to me. Do we expect any other package managers than "npm" or "yarn"? And if so, do we expect them to use `intall --frozen-lockfile` in that case?
How about raising an error to avoid proceeding with unknown results?

Reply:
iiuc, pants also support `pnpm`. I kept the trailing return to not break existing uses of this backend, I also tried to follow the pattern of all other properties returning a default. Maybe we could check if the package manager is supported prior to the use of this property?

Reply:
iiuc, pants also support `pnpm`. I kept the trailing return to not break existing uses of this backend, I also tried to follow the pattern of all other properties returning a default. Maybe we could check if the package manager is supported prior to the use of this property?

Reply:
Ah, right. OK, so I'd prefer it if this method would be explicit about it then (regardless of how it was before, I think this could be improved from status quo) by not silently accepting "any" backend.

i.e. something along the lines of:
```suggestion
    def immutable_install_args(self) -> tuple[str, ...]:
        if self.package_manager == "npm":
            return ("clean-install",)
        if self.package_manager == "yarn":
            if nodesemver.satisfies(self.package_manager_version, "1.x"):
                return ("install", "--frozen-lockfile")
            return ("install", "--immutable")
        if self.package_manager == "pnpm":
          return ("install", "--frozen-lockfile")
        raise E(f"Unsupported package manager: {self.package_manager}")
```

Reply:
Got it, changed to raise when an unsupported package manager is provided: https://github.com/pantsbuild/pants/pull/20540/commits/3a732f82978b25e39fc626aaef42c759c5b47eec.

Reply:
I think there's abstractly a bit of tension here with being Pants being overly constraining.

For instance, what if someone publishes a pnpm-CLI-compatible `cool-new-package-manager`  (similar to how `uv pip` is supposedly a drop-in replacement for `pip` https://github.com/astral-sh/uv)?

In theory, if we were less strict, we could have that drop in here and start working without needing pants changes... but, in practice, for this particular example, being strict sounds good: there's already 3 different styles of package manager, how will pants know that `cool-new-package-manager` should use pnpm vs. npm vs. yarn CLI? So yeah, exceptions plus dedicated pants support seems like the way forward 👍 

Taking it one step further, there seems to be a lot of `if self.package_manager == "<specific string>"` calls here. Maybe this `NodeJsProject` type should validate the `project_manager` as an enum, on construction, to give errors eagerly and not require each separate method to handle it. (I don't think we should block this targeted bugfix on making that improvement, though.)